README.md
# EMGdecomPy

[![ci-cd](https://github.com/UBC-SPL-MDS/emgdecompy/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-SPL-MDS/emgdecompy/actions/workflows/ci-cd.yml)
[![codecov](https://codecov.io/gh/The-Motor-Unit/EMGdecomPy/branch/main/graph/badge.svg?token=78ZU40UEOE)](https://codecov.io/gh/The-Motor-Unit/EMGdecomPy)

A package for decomposing multi-channel intramuscular and surface EMG signals into individual motor unit activity based off the blind source algorithm described in [`Negro et al. (2016)`](https://iopscience.iop.org/article/10.1088/1741-2560/13/2/026027/meta).

## Table of Contents

- [Overview](#overview)
- [Project Directory](#project-directory)
- [Proposal and Final Report](#proposal-and-final-report)
- [Installation](#installation)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)
- [Credits](#credits)

## Overview

### What's Been Accomplished

An open-source Python package, `EMGdecomPy` containing two elements, a blind source separation algorithm based on the work of [`Negro et al. (2016)`](https://iopscience.iop.org/article/10.1088/1741-2560/13/2/026027/meta) and a visualization element, has been created to decompose raw EMG signals into its constituent motor unit activity. Experimental durations of any length can be run using `EMGdecomPy`.

The blind source separation algorithm has been modified slightly from [`Negro et al. (2016)`](https://iopscience.iop.org/article/10.1088/1741-2560/13/2/026027/meta). The initialization process of the separation vectors has been changed so that instead of initializing every separation vector with the same time instance of highest activity in the pre-processed data, each subsequent vector is initialized with the next highest activity time instance in the pre-processed data.

More customization of the decomposition process is also allowed through different arguments to the `decomposition` function. For example, the separation vectors can be orthogonalized against each other using either the 'source deflation' process described in [`Negro et al. (2016)`](https://iopscience.iop.org/article/10.1088/1741-2560/13/2/026027/meta) or the Gram-Schmidt method.

We have not had the chance to thoroughly validate our algorithm but preliminary results look promising, as 3 out of 5 of the MUAP shapes identified by `EMGdecomPy` were also identified by [`Hug et al. (2021)`](https://figshare.com/articles/dataset/Analysis_of_motor_unit_spike_trains_estimated_from_high-density_surface_electromyography_is_highly_reliable_across_operators/13695937) for the **Gastrocnemius lateralis** muscle with 10% contraction intensity.  Refer to the [documentation](https://emgdecompy.readthedocs.io/en/stable/autoapi/emgdecompy/decomposition/index.html#emgdecompy.decomposition.decomposition) and the [final report](https://github.com/UBC-SPL-MDS/emg-decomPy/blob/main/docs/final-report/final-report.pdf) for more information.

The visualization element allows the user to interactively visualize the results of the blind source separation algorithm. The user can visualize one motor unit at a time from the motor units that were extracted from the EMG data using the algorithm. The visualization includes four plots, the instantaneous firing rate vs time, the signal vs time, an overlayed version of both the previous plots, and the average motor unit action potential shapes per channel. For a better idea of the interactivity of the plot, refer to the [`EMGdecomPy` workflow notebook](https://github.com/The-Motor-Unit/EMGdecomPy/blob/main/notebooks/emgdecompy-worfklow.ipynb).

### What's Not Working

Currently, the blind source separation algorithm accepts multiple motor units of the same shape. Upon inspection, it can be seen that many of these motor units have the exact same firing times or are time lagged from each other. Solutions to these problems are still in development, and include adding an orthogonalization step to the `refinement` function to stop the refinement process from converging on previous motor units and not accepting motor units whose firing times are within a certain time frame as another motor unit.

There is also a bug in the visualization component that does not allow the user to visualize the results of a decomposition if only one motor unit is accepted. This bug is due to how the peak shapes are created and a fix is currently in development.

### Future Work

Future work includes fixing the aforementioned problems, increasing code efficiency, improving the accuracy of the algorithm using domain knowledge, and further quantitative/qualitative validation of the results of the algorithm using the data from [`Hug et al. (2021)`](https://figshare.com/articles/dataset/Analysis_of_motor_unit_spike_trains_estimated_from_high-density_surface_electromyography_is_highly_reliable_across_operators/13695937) and other EMG data sources.

A further improvement to the algorithm would be a re-learning feature. The user would run the algorithm on a sample of the data, and then identify inaccurate firing times (false positives) based on physiological limits of motor unit firing rates. Then the algorithm would use this information to no longer make similar mistakes in the rest of the decomposition. Implementing this feature would be quite complex because it is algorithmically unclear how this would be done.

One idea is to somehow change the initialization of the separation vectors so that they no longer identify the false firing times when applied to the pre-processed data. However, since the separation vector changes throughout the LCA and refinement processes, it would be hard to control the effect that this would have on the estimated firing times. Another approach would be to influence the KMeans algorithm so that the threshold for the small peaks cluster includes the false positive peaks, in the hopes that future peaks of similar size are also false positives. The downside to this approach would be that we may increase the number of incorrect identifications of large peaks as small peaks, which are discarded.

An improvement to the visualization related to the above improvement would be the ability to remove peaks with a click of a button. This improvement is already in progress and if the re-learning feature is implemented then these two features can be connected.

## Project Directory

- [.githubworkflows](https://github.com/The-Motor-Unit/EMGdecomPy/tree/main/.github/workflows)
  - Contains file for automated testing and publishing of package.
- [data](https://github.com/The-Motor-Unit/EMGdecomPy/tree/main/data)
  - Contains a "raw" subdirectory with the EMG data corresponding to the **Gastrocnemius lateralis** muscle with 10% contraction intensity EMG data from [`Hug et al. (2021)`](https://figshare.com/articles/dataset/Analysis_of_motor_unit_spike_trains_estimated_from_high-density_surface_electromyography_is_highly_reliable_across_operators/13695937).
  - In the future can contain subdirectories pertaining to results from the blind source separation algorithm.
- [docs](https://github.com/The-Motor-Unit/EMGdecomPy/tree/main/docs)
  - Contains files related to the final report, the proposal, and the `ReadtheDocs` documentation.
- [notebooks](https://github.com/The-Motor-Unit/EMGdecomPy/tree/main/notebooks)
  - Contains a Jupyter notebook with the well-documented reproducible workflow that can be used to apply `EMGdecomPy` on EMG data and/or as a guide on how to use the package.
- [src](https://github.com/The-Motor-Unit/EMGdecomPy/tree/main/src/emgdecompy)
  - Contains the `.py` scripts containing `EMGdecomPy` source code.
- [tests](https://github.com/The-Motor-Unit/EMGdecomPy/tree/main/tests)
  - Contains the tests for the functions within `src`.

## Proposal and Final Report

To generate the proposal and final report locally, ensure that you have R version 4.1.2 or above installed, as well as the RStudio IDE. Then install the necessary dependencies with the following commands:

```
Rscript -e 'install.packages("rmarkdown", repos="http://cran.us.r-project.org")'
Rscript -e 'install.packages("tinytex", repos="http://cran.us.r-project.org")'
Rscript -e 'tinytex::install_tinytex()'
Rscript -e 'install.packages("bookdown", repos="http://cran.us.r-project.org")'
```

### Proposal

Our project proposal can be found [here](https://github.com/UBC-SPL-MDS/emg-decomPy/blob/main/docs/proposal/proposal.pdf).

To generate the proposal locally, run the following command from the root directory after cloning `EMGdecomPy`:

```
Rscript -e "rmarkdown::render('docs/proposal/proposal.Rmd')"
```

Alternatively, if the above doesn't work, install Docker. While Docker is running, run the following command from the root directory after cloning `EMGdecomPy`:

```

CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.6.7 (2022-06-29)


## v0.6.6 (2022-06-29)
### Documentation
* Change codecov settings ([`fe32513`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/fe325136106321cddae4519dd7ca6abf807a5fed))
* Change codecov badge to reflect new org name ([`41d5543`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/41d55433b62cc6a6cbb89d4b5acd83752eea012d))
* Get rid of docs badge - causing glitch on rtd ([`d0a65b4`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/d0a65b42d8cc500e54ae37637b950ecd285f003d))

## v0.6.5 (2022-06-29)
### Fix
* Just rerun notebook and comment in the decomposition ([`9763f79`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/9763f796832d646c13693f48671dd1085b0f0d0d))

## v0.6.4 (2022-06-29)
### Documentation
* Fix docs badge ([`edf26bd`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/edf26bd9804cd952f3267cedde839132285e3cef))
* Replace readthedocs latest branch with stable branch ([`5a1382b`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/5a1382b24a04758bdba80796745c5d0cf1b38f51))

## v0.6.3 (2022-06-29)


## v0.6.2 (2022-06-29)
### Documentation
* Use synonyms for affect and reword relearning part ([`f2ca67e`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/f2ca67e1e4c8ce1e3feeb63c5341d6fdcc037e2e))
* Make rscript commands copy-pasteable ([`e189d81`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/e189d811ad66c8903cdbaa6baf9e29ea429b96ac))
* Get rid of double inaccurate ([`5363fe5`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/5363fe52963aa807190a90aef8f31604440119c8))
* Standardize docstrings ([`cb4acf9`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/cb4acf9c1dc2716ed1ee7b2970eb10b593c917e8))
* Update docstrings ([`27d801a`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/27d801aea2f279b74cedd0a232e630c468d7c3ed))
* Add line about running notebook locally ([`2354f77`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/2354f7796cc7c11a408ff392e130f5b1fb38d51f))
* Fix toc links ([`3dc965b`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/3dc965b524be1531fb4d5de8bf84c817e98aa577))
* Add project directory: ([`24357e4`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/24357e4398accc82efb38104f5b68054a22afb6c))
* Add overview section ([`41d8f3c`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/41d8f3cf21b14eb06a03db20ff4edee983b4a8ca))
* Add copyright notice ([`06800c6`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/06800c651b5b243ac7127edb852ecacd3747eabf))
* Add toc ([`eeba7b4`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/eeba7b4bf486048f5269c0d68a0c4db0dd39a57c))

## v0.6.1 (2022-06-28)
### Documentation
* Further trim down report ([`f180279`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/f180279e2a7b34779a6629cccd9b1cdf2dad118d))

## v0.6.0 (2022-06-28)
### Feature
* Remove delete peak functions ([`7246973`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/7246973eea674379730423eeb922c17af03c0b2b))
* For the future. notebook but with remove peak working. ([`1fbd574`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/1fbd57483dce456bb8e3d4aa474d106e383ebc92))
* Add the remove false peak feature ([`52987ce`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/52987ce9b713c6497ec4d41a9059d25bdbc25a55))

### Fix
* Fix data compatibility issues with hug data and peak selection and optimize some code ([`31a83ce`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/31a83cecde6e1cc7e71964f1edfcc6a55a24f1a6))
* Fix RMSE broken ([`cdc976d`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/cdc976d7fd6713523820983416e0a30bb3d8ee1a))
* Latest notebook ([`64307b8`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/64307b89bb567dfd117d4ac701a7d8f68cbab5f6))
* Make dashboard nice and streamlined. widgets for everything. Remove Delete Peak from main notebook for now ([`8eeea97`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/8eeea97f488e02ab35f13ec4ad0ab9d3000e5d57))
* Method param is now str instead of function to allow for widget selection ([`e9b504a`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/e9b504a65143c051b75deb1cca116fd15a294669))
* Adding code to allow for team and source data to be visualised ([`bf821eb`](https://github.com/The-Motor-Unit/EMGdecomPy/commit/bf821eb74c8a0868d86460e951a7b82a2ed2ff6b))

src/emgdecompy/viz.py
from codecs import raw_unicode_escape_decode
import numpy as np
import pandas as pd
import altair as alt
import panel as pn
from panel.interact import interact, fixed
import math
from sklearn.metrics import mean_squared_error
from emgdecompy.preprocessing import (
    flatten_signal,
    center_matrix,
    butter_bandpass_filter,
)

pn.extension("vega")


def RMSE(arr1, arr2):
    """
    Evaluates Root Square Error for two series

    Parameters
    ----------
        arr1: iterative object
            First source, likely target.
        arr2: iterative object
            Second source, likely actual values.

    Returns
    -------
        float
            Root Mean Square Error of arr1 vs arr2.

    Examples
    --------
        >>> arr1 = [3, 4, 4, 9, 12]
        >>> arr2 = [3, 5, 3, 9, 11]
        >>> RMSE(arr1, arr2)
        0.7745966692414834

    """
    MSE = mean_squared_error(arr1, arr2)
    RMSE = math.sqrt(MSE)

    return RMSE


def mismatch_score(mu_data, peak_data, mu_index, method="RMSE", channel=-1):
    """
    Evaluates how well a given peak contributes to a given MUAP.
    This is called by muap_plot() function and is used to include error in the title of the muap plot.

    Parameters
    ----------
        mu_data: dict
            Dictionary containing MUAP shapes for each motor unit.
        peak_data: dict
            Dictionary containing shapes for a given peak per channel.
        mu_index: int
            Index of motor unit to examine
        method: function name
            Function to use for evaluating discrepency between mu_data and peak_data.
            Default: RMSE.
        channel: int
            Channel to run evaluation on.
            Defaul = -1 and it means average of all channels.

    Returns
    -------
        float
            Root Mean Square Error of MU data vs Peak data.
    """
    score = 0

    if channel == -1:  # For all channels, we can just
        # straight up compare RMSE across the board
        mu_sig = mu_data[f"mu_{mu_index}"]["signal"]
        peak_sig = peak_data[f"mu_{mu_index}"]["signal"]
        if method == "RMSE":
            score = RMSE(mu_sig, peak_sig)

    else:  # Otherwise, filter for a given channel
        # filter mu_data for signal data that channel
        indexes = np.where(mu_data[f"mu_{mu_index}"]["channel"] == channel)
        mu_sig = mu_data[f"mu_{mu_index}"]["signal"][indexes]

        indexes = np.where(peak_data[f"mu_{mu_index}"]["channel"] == channel)
        peak_sig = peak_data[f"mu_{mu_index}"]["signal"][indexes]
        if method == "RMSE":
            score = RMSE(mu_sig, peak_sig)

    return score


def muap_dict(raw, pt, l=31):
    """
    Returns multi-level dictionary containing sample number, average signal, and channel
    for each motor unit by averaging the peak shapes over every firing for each MUAP.

    Parameters
    ----------
        raw: numpy.ndarray
            Raw EMG signal.
        pt: numpy.ndarray
            Multi-dimensional array containing indices of firing times
            for each motor unit.
        l: int
            One half of action potential discharge time in samples.

    Returns
    -------
        dict
            Dictionary containing MUAP shapes for each motor unit.
    """
    raw = flatten_signal(raw)
    channels = raw.shape[0]
    shape_dict = {}
    pt = pt.squeeze()

    for i in range(pt.shape[0]):
        pt[i] = pt[i].squeeze()

        # Create array to contain indices of peak shapes
        ptl = np.zeros((pt[i].shape[0], l * 2 + 1), dtype="int")

        for j, k in enumerate(pt[i]):
            ptl[j] = np.arange(k - l, k + l + 1)

            if np.less(k, l) == True:
                ptl[j] = np.arange(k - l, k + l + 1)
                neg_idx = abs(k - l)
                ptl[j][:neg_idx] = np.repeat(0, neg_idx)

            else:
                ptl[j] = np.arange(k - l, k + l + 1)

        ptl = ptl.flatten()

        # Create channel index of each peak
        channel_index = np.repeat(np.arange(channels), l * 2 + 1)

        # Get sample number of each position along each peak
        sample = np.arange(l * 2 + 1)
        sample = np.tile(sample, channels)

        # Get average signals from each channel
        signal = (
            raw[:, ptl]
            .reshape(channels, ptl.shape[0] // (l * 2 + 1), l * 2 + 1)
            .mean(axis=1)
            .flatten()
        )

        shape_dict[f"mu_{i}"] = {
            "sample": sample,
            "signal": signal,
            "channel": channel_index,
        }

    return shape_dict


def muap_dict_by_peak(raw, peak, mu_index=0, l=31):
    """
    Returns the dictionary of shapes for a selected peak, by channel.
    It is called by the select_peak() function when a peak is selected by a user.

    Parameters
    ----------
        raw: numpy.ndarray
            Raw EMG signal.
        peak: int
            Peak timing to plot.
        mu_index: int
            Motor Unit the peak belongs to, to keep dict format consistent.
        l: int
            One half of action potential discharge time in samples.

    Returns
    -------
        dict
            Dictionary containing shapes for a given peak per channel.
    """
    raw = flatten_signal(raw)
    channels = raw.shape[0]
    shape_dict = {}

    low = peak - l
    high = peak + l + 1

    shape = raw[:, low:high]  # Shape is channels x Firings; frequently 63 x 62
    # Each of 62 values is a signal

    # Make dictionary from this data

    # Create channel index of each peak
    channel_index = np.repeat(np.arange(channels), l * 2 + 1)  # 64 zeros,
    # 64 ones,
    # 64 twos,
    # [...],
    # 64 sixty-threes.

    # Get sample number of each position along each peak
    sample = np.arange(l * 2 + 1)
    sample = np.tile(sample, channels)
    # sample <- [0,1,2,...,61,0,1,2,...,61]

    # Get signals of each peak
    signal = shape.flatten()
    if peak < l:
        neg_idx = abs(peak - l)
        signal[:neg_idx] = np.repeat(0, neg_idx)

    shape_dict[f"mu_{mu_index}"] = {
        "sample": sample,
        "signal": signal,
        "channel": channel_index,
    }

    return shape_dict


def channel_preset(preset="standard"):
    """
    Returns a dictionary with two keys:
    'sort_order' with the list to order channels,
    and 'cols' with the number of columns for a given channel arrangement.
    Called by muap_plot() function to determine the order of channels to plot shapes.

    Parameters
    ----------
    preset: str
        Name of the preset to use

    Returns
    -------
        dict with two keys
            cols: int
                Number of columns
            sort_order: list
                Sort order of all the channels

    Examples
    --------
        >>> channel_preset(preset='vert63')
        {
        'cols': 5,
        'sort_order': [
            63, 38, 37, 12, 11, 62, 39, 36, 13, 10, 61, 40, 35, 14, 9, 60, 41, 34, 15, 8, 59, 42, 33, 16, 7, 58, 43, 32, 17, 6, 57, 44, 31, 18, 5, 56, 45, 30, 19, 4, 55, 46, 29, 20, 3, 54, 47, 28, 21, 2, 53, 48, 27, 22, 1, 52, 49, 26, 23, 0, 51, 50, 25, 24
            ]
        }
    """

    if preset == "standard":
        sort_order = list(range(0, 64, 1))
        cols = 8

    elif preset == "vert63":
        sort_order = [
            63,
            38,
            37,
            12,
            11,
            62,
            39,
            36,
            13,
            10,
            61,
            40,
            35,
            14,
            9,
            60,
            41,
            34,
            15,
            8,
            59,
            42,
            33,
            16,
            7,
            58,
            43,
            32,
            17,
            6,
            57,
            44,
            31,
            18,
            5,
            56,
            45,
            30,
            19,
            4,
            55,
            46,
            29,
            20,
            3,
            54,
            47,
            28,
            21,
            2,
            53,
            48,
            27,
            22,
            1,
            52,
            49,
            26,
            23,
            0,
            51,
            50,
            25,
            24,
        ]
        cols = 5

    res = dict(cols=cols, sort_order=sort_order)

    return res


def muap_plot(
    mu_data, mu_index, peak_data=None, l=31, peak="", method="RMSE", preset="standard"
):
    """
    Returns a plot for MUAP shapes separated by channel.
    If peak_data is specified, also plots overlay of contribution of the peak to the shape per channel.
    Called by select_peak() function.

    Parameters
    ----------
        mu_data: dict
            Dictionary containing MUAP shapes for each motor unit.
        mu_index: int
            Index of motor unit to examine
        peak_data: dict
            Dictionary containing shapes for a given peak per channel.
            Specifying it creates the overlay of peak contribution
        l: int
            One half of action potential discharge time in samples.
        peak: int:
            Index of the peak, used for the Title of the plot.
        method: function name
            Function to use to calculate mean (over all channels) mismatch score between averaged shape and given peak
        preset: str
            Name of preset to use, for arranging the channels on the plot

    Returns
    -------
        altair.vegalite.v4.api.FacetChart
            Facetted altair plot overlaying MU shapes per channel
            and peak shapes per channel.
    """
    alt.data_transformers.disable_max_rows()

    df = pd.DataFrame(mu_data[f"mu_{mu_index}"])
    df["Source"] = "MUAP"
    plot_title = f"MUAP Shapes for MU {mu_index}"
    legend_position = None  # Hide legend when we only showing MUAPs
    sort_order = channel_preset(preset)["sort_order"]
    cols = channel_preset(preset)["cols"]

    if peak_data:
        peak_df = pd.DataFrame(peak_data[f"mu_{mu_index}"])
        peak_df["Source"] = "Peak Contribution"
        df = pd.concat([df, peak_df])
        err = mismatch_score(mu_data, peak_data, mu_index, method=method, channel=-1)
        err = round(err)
        plot_title = (
            f"Peak at {peak} s contribution per Channel to MU {mu_index}. RMSE = {err}"
        )
        legend_position = alt.Legend(
            orient="none",
            title=None,
            legendX=400,
            legendY=-40,
            direction="horizontal",
            titleAnchor="middle",
        )  # Show Legend when showing overlay

    selection = alt.selection_multi(fields=["Source"], bind="legend")

    plot = (
        alt.Chart(df, title=plot_title)
        .encode(
            x=alt.X("sample", axis=None),
            y=alt.Y("signal", axis=None),
            color=alt.Color(
                "Source",
                scale={"range": ["#fd3a4a", "#99a7f1"]},
                legend=legend_position,
            ),
            opacity=alt.condition(selection, alt.value(1), alt.value(0.2)),
            facet=alt.Facet(
                "channel",
                columns=cols,
                spacing={"row": 0},
                header=alt.Header(
                    titleFontSize=0,
                    labelFontSize=14,
                    # labelOrient="bottom",
                ),
                sort=sort_order,
            ),
        )
        .mark_line()
        .properties(width=112, height=100)
        .configure_title(fontSize=14, anchor="middle")
        .configure_axis(labelFontSize=14)
        .configure_view(strokeWidth=0)
        .add_selection(selection)
    )

    return plot


def pulse_plot(pt, c_sq_mean, mu_index, sel_type="single"):
    """
    Plot firings for a given motor unit.

    Parameters
    ----------
        pulse_train: np.array
            Pulse train.
        c_sq_mean: np.array
            Centered, squared and averaged firings over the duration of the trial.
        mu_index: int
            Motor Unit of interest to plot firings for.
            Default is None and means return all pulses.
        sel_type: str
            Whether to select single points or intervals

    Returns
    -------
        altair plot object
    """

    color_pulse = "#35d3da"
    color_rate = "#9cb806"

    mu_count = pt.squeeze().shape[0]

    motor_df = pd.DataFrame(columns=["Pulse", "Strength", "Motor Unit", "Hz"])

    for i in range(0, mu_count):
        # PT for MU of interest:
        pt_selected = pt.squeeze()[i].squeeze()
        strength_selected = c_sq_mean[pt_selected]
        hertz = np.insert(1 / np.diff(pt_selected) * 2048, 0, 0)

        # Make those into DF:
        pulses_i = {
            "Pulse": pt_selected,
            "Strength": strength_selected,
            "Motor Unit": i,
            "seconds": pt_selected / 2048,
            "Hz": hertz,
        }
        motor_df_i = pd.DataFrame(pulses_i)
        motor_df = pd.concat([motor_df, motor_df_i])

        motor_df = motor_df.loc[motor_df["Motor Unit"] == mu_index]
        # brush = alt.selection_interval(encodings=['x'], name='brush') # Don't know if we will use this

    # TODO: Selection only makes sense if we are working with specific MU

    sel_peak = alt.selection_single(name="sel_peak")

    sel_interval = alt.selection_interval(encodings=["x"], name="sel_interval")

    chart_top_base = (
        alt.Chart(motor_df)
        .encode(
            alt.X(
                "seconds:Q",
                axis=alt.Axis(title="Time (s)", grid=False),
            )
        )
        .properties(width=1000, height=100)
    )

    chart_top_rate = (
        chart_top_base.mark_point(size=30, color=color_rate)
        .encode(
            alt.Y(
                "Hz:Q",
                axis=alt.Axis(
                    title="Instantaneous Firing Rate (Hz)",
                    grid=False,
                    format=".0f",
                    titleColor=color_rate,
                ),
            )
        )
        .add_selection(sel_interval)
    )

    chart_top_pulse = chart_top_base.mark_bar(
        size=3.5, color=color_pulse, opacity=0.3
    ).encode(
        alt.Y(
            "Strength:Q",
            axis=alt.Axis(
                title="Signal (A.U.)", grid=False, format="s", titleColor=color_pulse
            ),
        )
    )

    chart_top = alt.layer(chart_top_pulse, chart_top_rate).resolve_scale(
        y="independent"
    )

    chart_rate = (
        alt.Chart(motor_df)
        .encode(
            alt.X(
                "seconds:Q",
                axis=alt.Axis(title="Time (s)", grid=False),
                scale=alt.Scale(domain=sel_interval),
            ),
            alt.Y(
                "Hz:Q",
                axis=alt.Axis(
                    title="Instantaneous Firing Rate (Hz)",
                    grid=False,
                    format=".0f",
                    titleColor=color_rate,
                ),
            ),
            color=alt.condition(
                sel_peak, alt.value(color_rate), alt.value("lightgray"), legend=None
            ),
            tooltip=[
                alt.Tooltip("Hz", format=".2f"),
                alt.Tooltip("seconds", format=".2f"),
            ],
        )
        .properties(width=1000, height=250)
        .mark_point(size=30)
        .add_selection(sel_peak)
        .transform_filter(sel_interval)
    )

    chart_pulse = (
        alt.Chart(motor_df)
        .encode(
            alt.X(
                "seconds:Q",
                axis=alt.Axis(title="Time (s)", grid=False),
                scale=alt.Scale(domain=sel_interval),
            ),
            alt.Y(
                "Strength:Q",
                axis=alt.Axis(
                    title="Signal (A.U.)",
                    grid=False,
                    format="s",
                    titleColor=color_pulse,
                ),
            ),
            color=alt.condition(
                sel_peak, alt.value(color_pulse), alt.value("lightgray"), legend=None
            ),
        )
        .mark_bar(size=3.5)
        .add_selection(sel_peak)
        .properties(width=1000, height=250)
        .transform_filter(sel_interval)
    )

    return chart_top & chart_rate & chart_pulse


def select_peak(
    selection, mu_index, raw, shape_dict, pt, preset="standard", method=RMSE
):
    """
    Retrieves a given peak (if any) and re-graphs MUAP plot via muap_plot() function.
    Called within dashboard() function, binded to the peak selection on pulse graphs.

    Parameters
    ----------
        selection: selection object
            Selection object to dig into and retrieve peak index to plot.

        mu_index: int
            Currently plotted Motor Unit.

        raw: numpy.ndarray
            Raw EMG signal array.

        shape_dict: dict
            Dictionary containing MUAP shapes for each motor unit.

        pt: numpy.ndarray
            Multi-dimensional array containing indices of firing times
            for each motor unit.

    Returns
    -------
        altair plot object

    """
    global selected_peak

    if not selection:
        plot = muap_plot(shape_dict, mu_index, l=31, preset=preset, method=RMSE)
        selected_peak = -1

    else:
        selected_peak = selection[0] - 1
        # for some reason beyond my grast these are 1-indexed
        peak = pt[mu_index][selected_peak]

        peak_data = muap_dict_by_peak(raw, peak, mu_index=mu_index, l=31)
        plot = muap_plot(
            shape_dict,
            mu_index,
            peak_data,
            l=31,
            peak=str(peak),
            preset=preset,
            method=RMSE,
        )

    return pn.Column(
        pn.Row(
            pn.pane.Vega(plot, debounce=10, width=750),
        )
    )


def dashboard(decomp_results, raw, mu_index=0, preset="standard", method="RMSE"):
    """
    Parent function for creating interactive visual component of decomposition.
    Dashboard consists of four plots:
    1. Plot of firings and signal, primarily for zooming and navigating.
    2. Plot of signal strength, which allows for peak selection.
    3. Plot of firings, which allows for peak selection.
    4. MUAP plot of individual motor unit shapes by channel, with selected peak overlay

    Parameters
    ----------
        decomp_results: dict
            Decomposition results.
            Must contain [MUPulses] key with the pulses array.

        raw: numpy.ndarray
            Raw EMG data.

        mu_index: int
            Currently plotted Motor Unit.

    Returns
    -------
        panel object containing interactive altair plots
    """

    signal = flatten_signal(raw)
    signal = np.apply_along_axis(
        butter_bandpass_filter,
        axis=1,
        arr=signal,
        lowcut=10,
        highcut=900,
        fs=2048,
        order=6,
    )
    centered = center_matrix(signal)
    c_sq = centered ** 2
    c_sq_mean = c_sq.mean(axis=0)

    pt = decomp_results["MUPulses"]

    shape_dict = muap_dict(raw, pt, l=31)
    pulse = pulse_plot(pt, c_sq_mean, mu_index, sel_type="interval")
    pulse_pn = pn.pane.Vega(pulse, debounce=10)
    mu_charts_pn = pn.bind(
        select_peak,
        pulse_pn.selection.param.sel_peak,
        mu_index,
        raw,
        shape_dict,
        pt,
        preset,
        method,
    )

    res = pn.Column(
        pulse_pn,
        mu_charts_pn,
    )

    return res


def b_click(event):
    """
    Function triggered by clicking "Delete Selected Peak" button on the dashboard
    Bound to the button widget inside dashboard() function
    Deletes selected peak from the output variable and reruns the dashboard

    Parameters
    ----------
        event: event
            event that triggered the funciton

    Returns
    -------
        Null
    """
    global selected_peak

    if selected_peak > -1:

        # Get the peak and the selected MU index
        ###############################
        peak = dash_p[1][0][0].object.data.iloc[selected_peak]["Pulse"]
        mu_index = dash_p[0][0].value

        # Change decomp_results:
        ###############################
        global output
        output = remove_false_peak(output, mu_index, peak)

        # Reconstruct the plot:
        ###############################
        raw = raw_data_dict["SIG"]
        decomp_results = output
        signal = flatten_signal(raw)
        signal = np.apply_along_axis(
            butter_bandpass_filter,
            axis=1,
            arr=signal,
            lowcut=10,
            highcut=900,
            fs=2048,
            order=6,
        )
        centered = center_matrix(signal)
        c_sq = centered ** 2
        c_sq_mean = c_sq.mean(axis=0)
        pt = decomp_results["MUPulses"]
        shape_dict = muap_dict(raw, pt, l=31)
        pulse = pulse_plot(pt, c_sq_mean, mu_index, sel_type="interval")
        pulse_pn = pn.pane.Vega(pulse, debounce=10)
        dash_p[1][0][0] = pulse_pn

        # Also redo mu_charts graph so that it no longer selects the deleted peak:
        mu_charts_pn = pn.bind(
            select_peak,
            pulse_pn.selection.param.sel_peak,
            mu_index,
            raw,
            shape_dict,
            pt,
            preset=preset,
            method=gl_method,
        )
        dash_p[1][0][1] = mu_charts_pn


def visualize_decomp(decomp_results, raw):
    """
    Wrapper function that allows for cleaner UI for user. Widgets are built within it.

    Parameters
    ----------
        decomp_results: dict
            Decomposition results.
            Must contain [MUPulses] key with the pulses array.

        raw: numpy.ndarray
            Raw EMG data.

    Returns
    -------
        panel object containing interactive altair plots
    """

    mu_index_widget = pn.widgets.Select(
        name="Motor Unit:",
        options=[i for i in range(0, len(decomp_results["MUPulses"]))],
        value=0,
    )
    mu_preset_widget = pn.widgets.Select(
        name="Preset:", options=["standard", "vert63"], value="standard"
    )
    mu_comp_widget = pn.widgets.Select(
        name="Comparison Metric:", options=["RMSE"], value="RMSE"
    )

    global dash_p

    dash_p = interact(
        dashboard,
        decomp_results=fixed(decomp_results),
        raw=fixed(raw),
        mu_index=mu_index_widget,
        preset=mu_preset_widget,
        method=mu_comp_widget,
    )

    # Build the button
    button_del = pn.widgets.Button(
        name="Delete Selected Peak", button_type="primary", width=50
    )

    button_del.on_click(b_click)

    res = pn.Column(
        button_del,
        dash_p,
    )

    return res
